hathor/consensus.py

```diff
@@ -30,20 +31,21 @@
 _base_transaction_log = logger.new()
 
 
+@dataclass
```

Comment:
As we have multiple fields and the `__init__()` method would have only two parameters, is it worth using a dataclass? I feel it's more complex with no benefits.

Reply:
These are basically the two versions:

Without dataclasses:

```python
class ConsensusAlgorithmContext:
    """ An instance of this class holds all the relevant information related to a single run of a consensus update.
    """

    consensus: 'ConsensusAlgorithm'
    pubsub: PubSubManager
    block_algorithm: 'BlockConsensusAlgorithm'
    transaction_algorithm: 'TransactionConsensusAlgorithm'
    txs_affected: Set[BaseTransaction] = field(default_factory=set)
    reorg_started: bool = False

    def __init__(self, consensus: 'ConsensusAlgorithm', pubsub: PubSubManager) -> None:
        self.consensus = consensus
        self.pubsub = pubsub
        self.block_algorithm = self.consensus.block_algorithm_factory(self)
        self.transaction_algorithm = self.consensus.transaction_algorithm_factory(self)
        self.txs_affected = set()
        self.reorg_started = False
```

With dataclasses:

```python
@dataclass
class ConsensusAlgorithmContext:
    """ An instance of this class holds all the relevant information related to a single run of a consensus update.
    """

    consensus: 'ConsensusAlgorithm'
    pubsub: PubSubManager
    block_algorithm: 'BlockConsensusAlgorithm' = field(init=False)
    transaction_algorithm: 'TransactionConsensusAlgorithm' = field(init=False)
    txs_affected: Set[BaseTransaction] = field(default_factory=set)
    reorg_started: bool = False

    def __post_init__(self) -> None:
        self.block_algorithm = self.consensus.block_algorithm_factory(self)
        self.transaction_algorithm = self.consensus.transaction_algorithm_factory(self)
```

I honestly much prefer the later, the side-effect being that it can now be constructed with more parameters (explicit txs_affected and reorg_started) which isn't useful, but I don't think it's harmful.

Reply:
After a discussion, we decided to keep the original code for now because instances are carrying more than data. We should refactor it later.